frontend/public/components/utils/resource-log.jsx

```diff
@@ -0,0 +1,249 @@
+
+    newState.logURL = this._logURL(props.model, namespace, objName, props.currentResource);
+    return newState;
+  }
+
+  // Returns API url for given resource log
+  _logURL(model, namespace, name, container = '') {
+    return resourceURL(model, {
+      ns: namespace,
+      name,
+      path: 'log',
+      queryParams: {
+        container: container,
+        follow: 'true',
+        tailLines: this._buffer.maxSize
+      }
+    });
+  }
+
+  // Updates log state as long as eof has not been reached
+  _updateLogState(newState) {
+
+    // If log is at eof, don't update state to anything other than eof
+    this.setState({
+      logState: this.props.eof ? 'eof' : newState,
+    });
+  }
+
+  // use _updateLogState toggle streaming/paused state instead of directly
+  // setting logState
+  _toggleStreaming() {
+    this._updateLogState(this.state.logState === 'streaming' ? 'paused' : 'streaming');
+  }
+
+  // Set loadtime
+  _touchLoadTime() {
+    this._touchLoadTimeState();
+    this._loadTime = Date.now();
+  }
+
+  // separate function so that it can be throttled
+  _touchLoadTimeState() {
+    this.setState({
+      loadTime: this.state.loadTime + 1
+    });
+  }
+
+  // Abort XMLHttpRequest and update logState accordingly
+  _endStreaming() {
+    // If paused, wait until next change in logState to set to eof
+    if (!this.logState === 'paused') {
+      this._updateLogState('eof');
+    }
+
+    this._stream.abort();
+    clearTimeout(this._pendingReload);
+  }
+
+  // Sets a timeout which will refresh the log from the API every five seconds,
+  // or after the previous load finishes, whichever comes first.
+  _resetPendingReload() {
+    const sinceLastLoad = Date.now() - this._loadTime;
+    const wait = Math.max(0, (1000 * 5) - sinceLastLoad);
+    this._pendingReload = setTimeout(() => {
+      if (this.state.logState === 'paused') {
+        // don't reset stream if the user paused the stream
+        this._touchLoadTime();
+        this._resetPendingReload();
+        return;
+      }
+      this._beginStreaming();
+    }, wait);
+  }
+
+  // Callback which handles the XMLHttpRequest "loadstart" event
+  _loadStarted() {
+    this._updateLogState('streaming');
+  }
+
+  // Callback to process data from the XMLHttpRequest "progress" event
+  _processData(data) {
+    if (dataHasHTML(data)) {
+      this._buffer.push('Logs are currently unavailable');
+    } else if (!dataHasFailureMsg(data)) {
+      this._buffer.push(data);
+    }
+    this._touchLoadTime();
+  }
+
+  // Use stream to load log data asynchronously
+  _beginStreaming() {
+    clearTimeout(this._pendingReload);
+    this._pendingReload = null;
+    this._buffer = lineBuffer(1000);
+
+    this._stream = stream(this.state.logURL, this._loadStarted.bind(this), this._processData.bind(this));
+    this._stream.promise
+      .then(() => {
+
+        // Resource is no longer running/generating new log content, so stop streaming
+        if (this.props.eof){
```

Comment:
Stop streaming if `eof` property is true. This means the container or build is no longer running.